airflow/providers/openlineage/extractors/dbapi_utils.py

```diff
@@ -0,0 +1,181 @@
+    hook: BaseHook,
+    namespace: str,
+    database: str | None,
+    in_query: str | None,
+    out_query: str | None,
+) -> tuple[list[Dataset], ...]:
+    """
+    This function queries database for table schemas using provided hook.
+    Responsibility to provide queries for this function is on particular extractors.
+    If query for input or output table isn't provided, the query is skipped.
+    """
+    in_datasets: list[Dataset] = []
+    out_datasets: list[Dataset] = []
+
+    # Do not query if we did not get both queries
+    if not in_query and not out_query:
+        return [], []
+
+    with closing(hook.get_conn()) as conn:
+        with closing(conn.cursor()) as cursor:
+            if in_query:
+                cursor.execute(in_query)
+                in_datasets += [x.to_dataset(namespace, database) for x in parse_query_result(cursor)]
+            if out_query:
+                cursor.execute(out_query)
+                out_datasets += [x.to_dataset(namespace, database) for x in parse_query_result(cursor)]
+    return in_datasets, out_datasets
+
+
+def parse_query_result(cursor) -> list[TableSchema]:
+    schemas: dict = {}
+    columns: dict = defaultdict(list)
+    for row in cursor.fetchall():
+        table_schema_name: str = row[_TABLE_SCHEMA]
+        table_name: str = row[_TABLE_NAME]
+        table_column: SchemaField = SchemaField(
+            name=row[_COLUMN_NAME],
+            type=row[_UDT_NAME],
+            description=None,
+        )
+        ordinal_position = row[_ORDINAL_POSITION]
+        try:
+            table_database = row[_TABLE_DATABASE]
+        except IndexError:
+            table_database = None
+
+        # Attempt to get table schema
+        table_key = ".".join(filter(None, [table_database, table_schema_name, table_name]))
+
+        schemas[table_key] = TableSchema(
+            table=table_name, schema=table_schema_name, database=table_database, fields=[]
+        )
+        columns[table_key].append((ordinal_position, table_column))
+
+    for schema in schemas.values():
+        table_key = ".".join(filter(None, [schema.database, schema.schema, schema.table]))
+        schema.fields = [x for _, x in sorted(columns[table_key])]
+
+    return list(schemas.values())
+
+
+def create_information_schema_query(
+    columns: list[str],
+    information_schema_table_name: str,
+    tables_hierarchy: TablesHierarchy,
+    uppercase_names: bool = False,
+    allow_trailing_semicolon: bool = True,
+) -> str:
+    """This function creates query for getting table schemas from information schema."""
+    sqls = []
+    for db, schema_mapping in tables_hierarchy.items():
+        filter_clauses = create_filter_clauses(schema_mapping, uppercase_names)
+        source = information_schema_table_name
+        if db:
+            source = f"{db.upper() if uppercase_names else db}." f"{source}"
+        sqls.append(f"SELECT {', '.join(columns)} " f"FROM {source} " f"WHERE {' OR '.join(filter_clauses)}")
```

Comment:
What if the column names contain spaces or other reserved characters -- this feels at least error prone if not directly a risk of SQLI vulns.

Reply:
In case of params: I could use `paramstyle` which is an attribute of `DbApiHook` but this still seems to be unreliable for many reasons.

Reply:
@ashb `columns` here aren't user provided columns, but names of columns in `information_schema`of database we want to query, which are passed by particular extractor. For example, the default is

```
    _information_schema_columns = [
        "table_schema",
        "table_name",
        "column_name",
        "ordinal_position",
        "udt_name",
    ]
```

and for Snowflake it is

```
    _information_schema_columns = [
        "table_schema",
        "table_name",
        "column_name",
        "ordinal_position",
        "data_type",
    ]
```

Added SqlAlchemy-like column quoting though.